File: src/lib/api/mofdb.ts
import type { PymatgenStructure } from '$lib/structure'
import { SERVER_URL } from './config'

function format_error_detail(detail: unknown): string {
  if (typeof detail === `string`) return detail
  if (Array.isArray(detail)) {
    return detail
      .map((d) => {
        if (typeof d === `object` && d?.msg) {
          const loc = Array.isArray(d.loc) ? d.loc.join(`.`) : ``
          return loc ? `${d.msg} (${loc})` : d.msg
        }
        return JSON.stringify(d)
      })
      .join(`; `)
  }
  return JSON.stringify(detail)
}

export interface MofHit {
  id: number
  name: string
  database: string
  elements: string[]
  n_elements: number
}

export interface MofSearchResult {
  hits: MofHit[]
  count: number
}

export interface MofStructureResult {
  structure: PymatgenStructure
  name: string
  database: string
}

export const MOFDB_DATABASES = [
  `CoREMOF 2019`,
  `CoREMOF 2014`,
  `CSD`,
  `hMOF`,
  `IZA`,
  `PCOD-syn`,
  `Tobacco`,
]

export async function searchMofs(
  body: { name?: string; database?: string; limit?: number },
  server_url = SERVER_URL,
): Promise<MofSearchResult> {
  const response = await fetch(`${server_url}/api/mofdb/search`, {
    method: `POST`,
    headers: { 'Content-Type': `application/json` },
    body: JSON.stringify({ limit: 50, ...body }),
  })
  if (!response.ok) {
    const err = await response.json().catch(() => ({ detail: response.statusText }))
    throw new Error(format_error_detail(err.detail) || `Server error: ${response.status}`)
  }
  return response.json()
}

export async function getMofStructure(
  name: string,
  database = ``,
  server_url = SERVER_URL,
): Promise<MofStructureResult> {
  const params = new URLSearchParams({ name })
  if (database) params.set(`database`, database)
  const response = await fetch(`${server_url}/api/mofdb/structure?${params.toString()}`)
  if (!response.ok) {
    const err = await response.json().catch(() => ({ detail: response.statusText }))
    throw new Error(format_error_detail(err.detail) || `Server error: ${response.status}`)
  }
  return response.json()
}

File: server/catgo/models/mofdb.py
"""Pydantic models for MOF-database (MOFX-DB) search."""

from pydantic import BaseModel, Field

from .structure import PymatgenStructure


class MofSearchRequest(BaseModel):
    name: str | None = Field(default=None, description="Name prefix search")
    database: str | None = Field(
        default=None,
        description="Source DB: CoREMOF 2014|CoREMOF 2019|CSD|hMOF|IZA|PCOD-syn|Tobacco",
    )
    limit: int = Field(default=50, ge=1, le=200)


class MofHit(BaseModel):
    # Round-trip key is (name, database). mofdb_client.fetch() has no `id` kwarg,
    # `mofid` is None in CoREMOF, and `name` alone is not unique (the same MOF is
    # mirrored across CoREMOF 2014/2019). `id` is the true unique int key (kept for
    # display / dedup) but cannot be queried back via fetch().
    id: int = 0
    name: str
    database: str = ""
    elements: list[str] = Field(default_factory=list)
    n_elements: int = 0


class MofSearchResult(BaseModel):
    hits: list[MofHit]
    count: int


class MofStructureResult(BaseModel):
    structure: PymatgenStructure
    name: str
    database: str
